subsys/bluetooth/mesh/dfd_srv.c

```diff
@@ -478,10 +531,71 @@ static int handle_upload_start_oob(struct bt_mesh_model *mod, struct bt_mesh_msg
 				   struct net_buf_simple *buf)
 {
 	struct bt_mesh_dfd_srv *srv = mod->user_data;
+	uint8_t uri_len;
+	uint8_t *uri;
+	uint16_t fwid_len;
+	uint8_t *fwid;
 
-	LOG_DBG("");
+	uri_len = net_buf_simple_pull_u8(buf);
+
+	if (uri_len > buf->len) {
+		return -EINVAL;
+	}
+
+	uri = net_buf_simple_pull_mem(buf, uri_len);
+	fwid_len = buf->len;
+	fwid = net_buf_simple_pull_mem(buf, fwid_len);
 
+	if (upload_is_busy(srv)) {
+#ifdef CONFIG_BT_MESH_DFD_SRV_OOB_UPLOAD
+		if (srv->upload.is_oob &&
+		    uri_len == srv->upload.oob.uri_len &&
+		    fwid_len == srv->upload.oob.current_fwid_len &&
+		    !memcmp(uri, srv->upload.oob.uri, uri_len) &&
+		    !memcmp(fwid, srv->upload.oob.current_fwid, fwid_len)) {
+			/* Same image, return SUCCESS for idempotency */
+			upload_status_rsp(srv, ctx, BT_MESH_DFD_SUCCESS);
+			return 0;
+		}
+#endif
+		upload_status_rsp(srv, ctx, BT_MESH_DFD_ERR_BUSY_WITH_UPLOAD);
+		return 0;
+	}
+
+#ifdef CONFIG_BT_MESH_DFD_SRV_OOB_UPLOAD
+	if (uri_len > CONFIG_BT_MESH_DFU_URI_MAXLEN ||
+	    fwid_len > CONFIG_BT_MESH_DFU_FWID_MAXLEN) {
+		upload_status_rsp(srv, ctx, BT_MESH_DFD_ERR_INTERNAL);
+		return 0;
+	}
+
+	struct bt_mesh_dfu_slot *slot = bt_mesh_dfu_slot_reserve();
+
+	if (slot == NULL) {
+		upload_status_rsp(srv, ctx, BT_MESH_DFD_ERR_INSUFFICIENT_RESOURCES);
+		return 0;
+	}
+
+	srv->upload.is_oob = true;
+	srv->upload.slot = slot;
+	memcpy(srv->upload.oob.uri, uri, uri_len);
+	srv->upload.oob.uri_len = uri_len;
+	memcpy(srv->upload.oob.current_fwid, fwid, fwid_len);
+	srv->upload.oob.current_fwid_len = fwid_len;
+	memcpy(&srv->upload.oob.ctx, ctx, sizeof(struct bt_mesh_msg_ctx));
+
+	int status = srv->cb->start_oob_upload(srv, srv->upload.slot, srv->upload.oob.uri,
+					       srv->upload.oob.uri_len,
+					       srv->upload.oob.current_fwid,
+					       srv->upload.oob.current_fwid_len);
+
+	if (status != BT_MESH_DFD_SUCCESS) {
+		upload_status_rsp(srv, ctx, status);
+		bt_mesh_dfu_slot_release(srv->upload.slot);
+	}
+#else
 	upload_status_rsp(srv, ctx, BT_MESH_DFD_ERR_URI_NOT_SUPPORTED);
```

Comment:
This is a bit unclear for me. If a node doesn't support OOB (Out-of-band Retrieval Support state is 0) Why can't we just return immediately?

Reply:
Well, the spec says that supporting the message is M (mandatory) in 6.2.1, and the entire description for Receiving a Firmware Distribution OOB Start message does not mention that you can just not respond if you set that support state to 0. The way I read it (and the way it was before I started this entire OOB work), we have to respond. Responding with URI_NOT_SUPPORTED seems to be the most sensible status.

Reply:
I meant that we will call `upload_is_busy` even if we know we don't support OOB. And this code will stay there. We do we need to keep this code if we know we can return (return message) earlier.

Reply:
Yes, that is how I read the spec. If the upload is busy (with in-band, in this case) and you receive an OOB start message, you must return BUSY status. The only allowable status responses when you are halfway through an in-band upload for this message are BUSY and SUCCESS (and, of course, INTERNAL_ERROR)